warehouse/packaging/tasks.py

```diff
@@ -111,3 +113,54 @@ def update_description_html(request):
     for description in descriptions:
         description.html = readme.render(description.raw, description.content_type)
         description.rendered_by = renderer_version
+
+
+@tasks.task(
+    bind=True,
+    ignore_result=True,
+    acks_late=True,
+    autoretry_for=(Exception,),
+    retry_backoff=15,
+    retry_jitter=False,
+    max_retries=5,
+)
+def upload_bigquery_distributions(task, request, file, form):
+    """
+    Adds release file metadata to public BigQuery database
+    """
+    bq = request.find_service(name="gcloud.bigquery")
+
+    # Using the schema to populate the data allows us to automatically
+    # set the values to their respective fields rather than assigning
+    # values individually
+    def populate_data_using_schema(schema, row_data):
+
+        for sch in schema:
+            field_data = None
+
+            # The order of data extraction below is determined based on the
+            # classes that are most recently updated
+            if hasattr(file, sch.name):
+                field_data = getattr(file, sch.name)
+            else:
+                field_data = form[sch.name].data
+
+            if isinstance(field_data, datetime.datetime):
+                field_data = field_data.isoformat()
+
+            # Replace all empty objects to None will ensure
+            # proper checks if a field is nullable or not
+            if not isinstance(field_data, bool) and not field_data:
+                field_data = None
+
+            if field_data is None and sch.mode == "REPEATED":
+                row_data[sch.name] = []
+            else:
+                row_data[sch.name] = field_data
+
+    distribution_row_data = dict()
+    table_name = request.registry.settings["warehouse.distribution_table"]
+    populate_data_using_schema(bq.get_table(table_name).schema, distribution_row_data)
```

Comment:
Do we actually need to do `bq.get_table(table_name).schema` here every time? Or can we keep some representation of this in the codebase instead?

Reply:
In addition, I'd recommend not passing in a variable to be mutated here, making it the return value instead, and passing in the file/form as arguments. Something like:

```
json_rows = [file_and_form_to_row(file, form)]
```

Reply:
> Do we actually need to do `bq.get_table(table_name).schema` here every time? Or can we keep some representation of this in the codebase instead?

The reason for querying the schema every time is so that if the schema changes in the future or if a new table is created, we would only need to change it in BigQuery rather than also in the codebase. If query cost is an issue, I believe it is very minimal compared to the `insert_rows_json` command

Reply:
OK, that makes sense. Comment still stands about refactoring this a bit, though.